Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicHandler.cpp

```diff
@@ -206,34 +229,25 @@ namespace AzToolsFramework
                         "(A null instance is returned)."));
                 }
 
-                containerEntityId = instanceToCreate->get().GetContainerEntityId();
+                newContainerEntityId = instanceToCreate->get().GetContainerEntityId();
+                const TemplateId newInstanceTemplateId = instanceToCreate->get().GetTemplateId();
 
                 // Apply the correct transform to the container for the new instance, and store the patch for use when creating the link.
-                PrefabDom patch = ApplyContainerTransformAndGeneratePatch(containerEntityId, commonRootEntityId,
+                PrefabDom linkPatch = ApplyContainerTransformAndGeneratePatch(newContainerEntityId, commonRootEntityId,
                     containerEntityTranslation, containerEntityRotation);
 
-                // Update the cache - this prevents these changes from being stored in the regular undo/redo nodes
-                m_prefabUndoCache.UpdateCache(containerEntityId);
-
-                // Parent the non-container top level entities to the container entity.
+                // Parent the non-container top level entities to the new container entity.
                 // Parenting the top level container entities will be done during the creation of links.
                 EditorEntityContextNotificationBus::Broadcast(&EditorEntityContextNotification::SetForceAddEntitiesToBackFlag, true);
                 for (AZ::Entity* topLevelEntity : topLevelEntities)
                 {
                     if (!IsInstanceContainerEntity(topLevelEntity->GetId()))
                     {
-                        AZ::TransformBus::Event(topLevelEntity->GetId(), &AZ::TransformBus::Events::SetParent, containerEntityId);
+                        AZ::TransformBus::Event(topLevelEntity->GetId(), &AZ::TransformBus::Events::SetParent, newContainerEntityId);
                     }
                 }
                 EditorEntityContextNotificationBus::Broadcast(&EditorEntityContextNotification::SetForceAddEntitiesToBackFlag, false);
 
-                // Update the template of the instance since the entities are modified since the template creation.
-                Prefab::PrefabDom serializedInstance;
-                if (m_instanceToTemplateInterface->GenerateDomForInstance(serializedInstance, instanceToCreate->get()))
-                {
-                    m_prefabSystemComponentInterface->UpdatePrefabTemplate(instanceToCreate->get().GetTemplateId(), serializedInstance);
-                }
-
```

Comment:
This needs to be added back because this was capturing the parenting information of top level entities and updating the template with that information. This still needs to happen and it needs to happen before adding in nested prefabs too.

I'm surprised that this is not wrapped in undo-redo logic. You can consider adding that too.

Reply:
```
// Example
SuperCar.prefab      (being edited)
|-- Car.prefab       (been created)
    |-- Engine
        |-- Piston
    |-- Wheel.prefab
```

> This needs to be added back because this was capturing the parenting information of top level entities and updating the template with that information. This still needs to happen and it needs to happen before adding in nested prefabs too.

Yes. I took a close look. This update currently does two things in the template:
1. Update Car's container entity. After `prefabSystemComponent->CreatePrefab()`, the container entity DOM has 0 component. So after this step, it has all components in place and the child sort component has top level entities in the array.
2. Update transform component of top level entities (e.g. Engine) with new parent info

So, I suggest we keep this original code to do both work at once rather than selectively doing update for each.

Reply:
> I'm surprised that this is not wrapped in undo-redo logic. You can consider adding that too.

After the discussion, I thought about it and it seems like no undo-redo logic for the newly created template is expected and right behavior.

Imagine if we add undo-redo support for it and we hit undo after prefab creation. The Car template will be in a state with no entities and instances, or it can be worse if the template is in a weird state with some stuff but it is not complete (e.g. missing parent/child info). What happens if at this time we try to instantiate a Car prefab? I assume the Car template needs not be reloaded from disk because we already have the in-memory representation. We didn't clear the Car template object during undo.

***If there is no redo/undo support for Car, how redo works to put back the Car prefab instance?***

For the whole Car instance DOM, it can be redo because the link creation between SuperCar and Car supports redo/undo.

***If there is no redo/undo support for Car, how undo/redo works for top level entities (e.g. Engine)?***

For top level entities, their parent data in transform component is stored in SuperCar template "Entities". It supports the undo operation. For redo,  Car template is valid and complete and it contains the parent info in top level entity DOM. I think the redo of Car instance DOM is captured in the link creation between SuperCar and Car.

Reply:
In the new revision, I put it back. So it worked like before. It currently does two things as mentioned above: 
1. Update container entity DOM with all the components
2. Update top level entity DOMs and non-top level entity DOMs

It looks good for now. Later when we address the sort issue, we can change it to be more selective.